### PR TITLE
C51-344: Add date range filter to the  Activity feed

### PR DESCRIPTION
**Pull request by AkA84**
This PR adds the ability to filter the activities in the Activity Feed by date, offering a "range" selector (from...to)

range-filter

### Adding the filter
Strangely enough, the `civicaseActivityFilters` directive was already supporting a date range filter, as shown by its template
```html
<span ng-repeat="field in availableFilters" ng-switch="field.html_type">
  <span ng-switch-when="Select Date">
    <civicase-ui-date-range placeholder="{{ field.label }}" date-range="filters[field.name]"></civicase-ui-date-range>
  </span>
</span>
```
The filter was simply not present in `$scope.availableFilters` (stranger still was that  I could not find a point in the commit history where the filter was in `availableFilters` at all), so all that was needed was to add it
```js
function prepareAvailableFilters () {
  var availableFilters = [
    // ...        
    {
      name: 'activity_date_time',
      label: ts('Activity date'),
      html_type: 'Select Date'
    }
  ];
}
```
### Enforcing a time to the selected dates
The `civicaseUiDateRange` is simply a wrapper that contains two `crm-ui-datepicker` directives, that are configured to hide the time picker field (`time: false`), and it's implemented both in the Activity and Case filters
```html
<div class="civicase__ui-range">
  <input class="form-control" crm-ui-datepicker="{time: false}" ng-model="input.from" ng-placeholder="ts('From')"/>
  <input class="form-control" crm-ui-datepicker="{time: false}" ng-model="input.to" ng-placeholder="ts('To')"/>
</div>
```
The fact that the datepickers are configured to hide the time works fine when filtering cases, given that their `start_date` and `end_date` fields are of type `DATE`, but it doesn't when it comes to activities, given that the `activity_date_time` field is of type `DATETIME`.

If used as-is, the directive would have been unable for example to find all the activities of a specific day, say 2018-07-01, given that the resulting query param
```js
{
  activity_date_time: {
    BETWEEN: [ '2018-07-01', '2018-07-01' ]
  }
}
```
makes the api endpoint ignoring activities with `activity_date_time` with values like `2018-07-01 12:39:30`, `2018-07-01 10:00:00` etc

To solve the problem the directive was augmented with the `enforce-time` attribute
```html
<civicase-ui-date-range enforce-time placeholder="{{ field.label }}" date-range="filters[field.name]"></civicase-ui-date-range>
```
which makes the directive enforcing the `00:00:00` time for the "from" date, and the `23:59:59` time for the "to" date.

Given that this behaviour needs to be opted-in via `enforce-time` the directive can be still safely used for filtering cases as well

### CODE REVIEW

**Comment by jamienovick** [comment]
should we use CiviCRM's relative dates? i.e. last week / month etc:
see equivalent on advanced search: http://nimb.ws/M1lt75

**Comment by AkA84** [comment]
@jamienovick i've looked into it and it doesn't seems likely: the [logic](https://github.com/civicrm/civicrm-core/blob/master/CRM/Utils/Date.php#L1117-L1228) that handles the different range values is used primarily in the backend code of most (if not all) of the search forms (see [here](https://github.com/civicrm/civicrm-core/search?p=2&q=%3A%3AconvertFormValues&unscoped_q=%3A%3AconvertFormValues), when searching for `convertFormValues` which makes use of the aforementioned logic)

They seems to have [exposed](https://github.com/civicrm/civicrm-core/blob/master/api/v3/utils.php#L501) that logic also to be used in the api but unfortunately from the look of it it was more of an afterthought, and done only for [two entities](https://github.com/civicrm/civicrm-core/search?q=_civicrm_api3_get_using_query_object&unscoped_q=_civicrm_api3_get_using_query_object): Contact and Contribution

Not to mention that it seems we don't have that dropdown list of date ranges available to be used in an Angular app, as that list is being generated on the backend (at least it's the case for the [Activities search form](https://github.com/civicrm/civicrm-core/blob/master/CRM/Activity/BAO/Query.php#L446))

File: ang/civicase/Utils.js [hunk 1]
```diff
@@ -443,27 +443,61 @@
...
+      var enforceTime = attrs.hasOwnProperty('enforceTime');
+
       // Respond to user interaction with the date widgets
       element.on('change', function (e, context) {
         if (context === 'userInput' || context === 'crmClear') {
           $timeout(function () {
             if ($scope.input.from && $scope.input.to) {
-              $scope.data = {BETWEEN: [$scope.input.from, $scope.input.to]};
+              $scope.data = { BETWEEN: [
+                setAsRangeLimit($scope.input.from, 'lower'),
```
Comment (reneolivo) [inline ang/civicase/Utils.js]: why not `moment($scope.input.from).startOf('day').format(),` etc?

Comment (reneolivo) [reply ang/civicase/Utils.js]: The full example:

```js
if ($scope.input.from && $scope.input.to) {
  $scope.data = { BETWEEN: [
    moment($scope.input.from).startOf('day').format(),
    moment($scope.input.to).endOf('day').format()
  ] };
} else if ($scope.input.from) {
  $scope.data = moment($scope.input.from).startOf('day').format();
} else if ($scope.input.to) {
  $scope.data = moment($scope.input.to).endOf('day').format();
} else {
  $scope.data = null;
}
```

Comment (reneolivo) [reply ang/civicase/Utils.js]: Sorry, this solution does not work with `enforceTime`

**Comment by jamienovick** [comment]
@aka84 ok. @shitijg please note this for a future improvement

**Pull request merged**